### ISSUE TITLE
Wildcard expansion is silently disabled when a wildcard includes a file over MAX_PATH

### ISSUE THREAD

**Issue by rainersigwald** [Microsoft/msbuild#406]
This was reported as [a Connect issue](http://connect.microsoft.com/VisualStudio/feedback/details/2116449/msbuild-fails-to-find-files-in-directory-with-huge-sub-directory).

The repro is simple:
```xml
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <ItemGroup>
        <MyItem Include="MyDir\**\*.*" Exclude="MyDir\node_modules\**\*.*;MyDir\tmp\**\*.*" />
    </ItemGroup>

    <Target Name="TestInputs">
        <Warning Text="Inputs = @(MyItem)" />
    </Target>
</Project>
```
 
Where there are files in `MyDir\node_modules` that exceed `MAX_PATH`.  That returns:
```
C:\Users\raines\Downloads>msbuild test.proj
Microsoft (R) Build Engine version 14.0.24720.0
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 12/16/2015 12:46:15 PM.
Project "C:\Users\raines\Downloads\test.proj" on node 1 (default targets).
C:\Users\raines\Downloads\test.proj(8,3): warning : Inputs = MyDir\**\*.*
Done Building Project "C:\Users\raines\Downloads\test.proj" (default targets).

Build succeeded.
```

Note that the wildcards weren't actually expanded.

The original issue mentions an expectation that the `Exclude` element would prevent this from happening because that directory and all its contents are excluded. The reason that doesn't work is because we build the `Include` list fully first, then build the `Exclude` list, then do a subtraction. The failure occurs when building the `Include` list, so `Exclude` has no effect.

**Comment by rainersigwald** [comment]
There are a couple of questions here.

1. We have (until we figure out #53 for full-framework MSBuild) nice errors that we are supposed to emit when something goes over `MAX_PATH`.  Why didn't they fire here?
2. Should we precompute the impact of the exclude list as this user seems to expect?

For 2, I think the answer is "no". Being able to do pattern-based set exclusions seems like an awful lot of work for fairly minimal gain.

**Comment by shift-evgeny** [comment]
Thanks for troubleshooting this! You're right about my expectations. At minimum MSBuild should report a meaningful error here, but yes, I do expect that excluding the directory will avoid any problems with it.

"Precomputing the impact of the exclude list" is an implementation detail. From the user point of view, the ItemList is "everything inMyDir, except node_modules" - same as manually listing every other sub-directory in the include list (which works).

Yes, it is a problem with NPM that it builds a directory structure that exceeds MAX_PATH, but I have no control over that. All I can do is exclude the directory.

**Comment by rainersigwald** [comment]
Thinking about it a bit more, it may not be that difficult to consider the `Exclude` list inline with `Include`, but I haven't looked at that corner of the code in a while. If it's easy we should do it.

If it's not easy, I think fixing the long-path behavior would be sufficient to avoid this problem--it's inefficient to add and then remove things from a list but would produce the desired behavior.

**Comment by rainersigwald** [comment]
@shift-evgeny `Exclude` should now behave as you expect, thanks to improvements made for lazy item evaluation (#770). MSBuild now considers the exclude list while recursing through the include list and will stop when it hits an excluded directory.

This should already be available in VS "15" Preview 5.

I'm not closing this issue because the bad doesn't-expand-wildcards behavior persists if you _don't_ have an exclude. But at least the obvious workaround for `node_modules` works now.

**Comment by Daniel15** [comment]
The built-in `EmbeddedResource` stuff totally breaks because of this because it uses a wildcard like `**/*.resx` somewhere, resulting in the `GenerateResource` task throwing this error:
```
Microsoft.Common.CurrentVersion.targets(2867,5): error MSB3552: Resource file "**/*.resx" cannot be found.
```

https://github.com/dotnet/cli/issues/6561

If it's difficult to properly fix this, could you please at least update the `resx` stuff in `Microsoft.Common.CurrentVersion.targets` so it doesn't fail due to this problem?

**Comment by RehanSaeed** [comment]
Just started getting this error from Travis CI [here](https://travis-ci.org/RehanSaeed/Schema.NET/jobs/249300823). My path doesn't even seem that long. 

> /usr/share/dotnet/sdk/1.0.1/Microsoft.Common.CurrentVersion.targets(2865,5): error MSB3552: Resource file "**/*.resx" cannot be found. [/home/travis/build/RehanSaeed/Schema.NET/Source/Schema.NET/Schema.NET.csproj]

I guess my solution is to somehow tell Travis CI to put my code in folder at a lower level.

**Comment by Daniel15** [comment]
One cause for that error in web apps is node_modules. Older versions of npm
had very deep directory structures. If you're using npm, you should ensure
that it's the latest version rather than an older version like npm2. You
can run "npm install --global npm" to upgrade it.

Sent from my phone.

On Jul 2, 2017 9:43 AM, "Muhammad Rehan Saeed" <notifications@github.com>
wrote:

Just started getting this error from Travis CI here
<https://travis-ci.org/RehanSaeed/Schema.NET/jobs/249300823>. My path
doesn't even seem that long.

/usr/share/dotnet/sdk/1.0.1/Microsoft.Common.CurrentVersion.targets(2865,5):
error MSB3552: Resource file "**/*.resx" cannot be found.
[/home/travis/build/RehanSaeed/Schema.NET/Source/
Schema.NET/Schema.NET.csproj]

I guess my solution is to somehow tell Travis CI to put my code in folder
at a lower level.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/Microsoft/msbuild/issues/406#issuecomment-312502798>,
or mute the thread

**Comment by RehanSaeed** [comment]
Not using npm, just a simple class library.

**Comment by Daniel15** [comment]
Try running `find .` in your build script (which will recursively list all the files in the directory) and see if there's any very long paths. A long path **anywhere** in the directory will break the wildcard expansion.

**Comment by RehanSaeed** [comment]
@Daniel15 Thanks for the suggestion. Just tried that and this is what I found:

The longest path from the project causing the build error is **110 characters**:
> /home/travis/build/RehanSaeed/Schema.NET/Source/Schema.NET/health-lifesci/MedicalGuidelineContraindication.cs

The longest path from another project in the solution is **118 characters**:
> /home/travis/build/RehanSaeed/Schema.NET/Source/Schema.NET.Tool/Overrides/AddNumberTypeToMediaObjectHeightAndWidth.cs

These paths are well short of the 260 character max path, unless I'm missing something about Linux or MacOS where I'm running the `dotnet build` command.

**Comment by cellvia** [comment]
i am totally blocked by this.  I am also on osx and having the same results as @RehanSaeed 
looks like msbuild / dotnet core are not actually ready for unix :-/

**Comment by cellvia** [comment]
@RehanSaeed   for what its worth,  it seems like some issue with the hidden keys folder thats generated when launching debug session.  

to solve this: i created a keys folder (not hidden), and then within that a file with the same filename as the one that had been generated in the hidden .keys folder.  i left the file empty.

now every time i build / launch it works fine...

**Comment by rashidovich** [comment]
VM Ubuntu 17.04
.net core 2.0

dotnet build throws same issue.

basically i cannot deploy to production.
on local machine (OSX) builds fine.

i don't see any '.keys' folder.

Thanks in advance.

**Comment by dasjestyr** [comment]
Having same issue on both osx and ubuntu. Getting the latest version of npm did not help. Moving the project to root in order to shorten the path also did not work. No keys file either. Won't build locally or as a docker build on osx or ubuntu, but builds fine in a windows environment both locally and as docker build.

**Comment by radical** [comment]
Can you try running `find . -name '*\\*'` in your project directory?

**Comment by dasjestyr** [comment]
@radical not sure if you were talking to me, but if I run it, it just returns a single file which is an NServiceBus log file.

**Comment by radical** [comment]
@dasjestyr I hit this issue too and have a workaround. The problem is that the wildcard expansion is skipped if there is any error while trying to expand the wildcard. And the `\` in the path here gets converted to `/` on !windows and that causes a failure later. My workaround really just skips such a directory, and so doesn't really solve the issue. https://gist.github.com/radical/27ab744e235badd5f7b24dbfc85ec9af 
I will try to look at a proper solution later. But anyone else should feel free to take a stab at it meanwhile :)

**Comment by radical** [comment]
@dasjestyr um btw, was it a file or a directory for you? My patch might need to be fixed to handle files too. But this is a hack anyway.

**Comment by dasjestyr** [comment]
It returned: "./DAS.Gateways.Services.Api/C:\Logs\DAS\DAS.Infrastructure\log-{Date}.log"

**Comment by radical** [comment]
@dasjestyr It should be breaking only for a directory. Could you run`$ file ./DAS.Gateways.Services.Api/C:\Logs\DAS\DAS.Infrastructure\log-{Date}.log`?

**Comment by dasjestyr** [comment]
> ./DAS.Gateways.Services.Api/C:LogsDASDAS.Infrastructurelog-{Date}.log: cannot open `./DAS.Gateways.Services.Api/C:LogsDASDAS.Infrastructurelog-{Date}.log' (No such file or directory)

**Comment by radical** [comment]
You missed the `\`es in the path. Just check if `C:\Logs\DAS\DAS.Infrastructure\log-{Date}.log` is a directory or a file. I just want to confirm that I have the correct issue and we don't need to dig more.

**Comment by dasjestyr** [comment]
Oh, it's a directory

**Comment by olehdutko** [comment]
@cellvia , thanks for solution of fixing this issue!
Where I need to create  keys folder?

**Comment by DooMachine** [comment]
Deleting `<EnableDefaultContentItems>false</EnableDefaultContentItems>` in .csproj worked for me.

**Comment by andersforsgren** [comment]
The globbing seems to fail if any part of the traversal fails. In my case the problem was that the project output directory contained an NTFS junction that wasn't traversable, giving the same error as if it was traversable but longer than MAX_PATH. So being over MAX_PATH is not the only problem that gives these symptoms.

This is probably hard to fix for the general case with MAX_PATH, but it would be great if the globbing could report sensible errors, or even revover from some problems (e.g diagnostics that describe it did not glob one subdirectory because access was denied or a path was too long)

**Comment by markbearden** [comment]
I have a variation on this which might explain some of the reports such as failing when each file path is well under 200 characters.  (See RehanSaeed's Jul 5, 2017 comment, for example.)

I found this issue after having the following symptoms, which are abstracted here in order to make this easier to follow:

This kind of wild-card pattern was failing.  The failure was silent (no error stopped the build).
```
  <Target Name="MyTarget" DependsOnTargets="OtherTargets">
    <ItemGroup>
        <EmbeddedResource Include="Folder1\**\*.resx" />
    </ItemGroup>
  </Target>
```
The wild card expansion included roughly matching 30 file paths.  But none of the expanded paths for any individual file were even as long as 125 characters.   This seemed to indicate it's not the length of any *single* given file's path.

I experimentally shortened the path length of the build location, and found that below a certain length, the case above started working.  This confirms that path length is a factor.

After lots of experimenting, I found that THIS pattern worked even with my original longer paths, where the same set of files was covered by the union of the three ItemGroup elements, as was previously covered by the single ItemGroup element.   (These three ItemGroup wild cards together match the same set of ~30 files.)  This leads me to **suspect that the limitation is not due to a MAXLENGTH of any single file path, but to the total length of all the expanded file paths (presumably concatenated into a single temporary fixed-size buffer).**  In other words, **it's a combination of path length (including folder depth) AND the number of elements matching the wild card expansions**?
```
  <Target Name="MyTarget" DependsOnTargets="OtherTargets">
    <ItemGroup>
        <EmbeddedResource Include="Folder1\*\*.resx" />
    </ItemGroup>
      <ItemGroup>
        <EmbeddedResource Include="Folder1\Folder2\*\*.resx" />
      </ItemGroup>
      <ItemGroup>
        <EmbeddedResource Include="Folder1\Folder2\Folder3\*.resx" />
      </ItemGroup>
  </Target>
```

**Comment by ghidalgo3** [comment]
If MSBuild can detect when an item group evaluation has reached MAX_PATH, at least log a normal or diagnostic log saying "ItemGroup evaluation for X reached MAX_PATH at C:\longpath... and will be discarded" so that the user sees this is happening to them.

We ran into this issue internally due to a misbehaving program that generated a very long path which then causes MSBuild to suddenly stop building inexplicably because it couldn't expand a dirs.proj with an `<ProjectReferences Include="src/**/*.csproj">`. I wish I could get back a day of trying to figure why my build suddenly broke :)

**Comment by ghidalgo3** [comment]
@rainersigwald Would you be opposed to adding diagnostic logs when this situation is detected? I can make the change and raise a PR if you don't have any objections and if it would not be a breaking change.

**Comment by rainersigwald** [comment]
@ghidalgo3 No, that sounds good. But I'm not sure the logging context is available at all the right places. Feel free to take a shot at it and put up a draft PR.

A message wouldn't be a breaking change, but we'll have to balance chattiness/usefulness. For MAX_PATH it's likely always nice. For "couldn't expand and falling back to literal" in general, a lot of times that's intentional and we shouldn't always message.

**Comment by mjsabby** [comment]
Let's reboot this issue given there is support for MAX PATH now in .NET

cc @jaredpar

**Comment by jaredpar** [comment]
Curious: this issue seems to predate us having pretty solid long path support in .NET. Now though it's a fairly established feature and supported by our build tools. Given the current state of our tech is there a reason why this still needs to be unsupported? Basically can wildcards now expand beyond `MAX_PATH`? Or is there some limitation here that stops that even with us having good long path support?

**Comment by rainersigwald** [comment]
@mjsabby what made you comment on this? Exceeding MAX_PATH in a glob should not hit this problem any more (in a context where long paths are supported, i.e. on modern OS with long paths enabled and anywhere outside of Visual Studio's `devenv.exe`).

**Comment by rainersigwald** [comment]
> Basically can wildcards now expand beyond `MAX_PATH`?

Yes, in any case where the long-path support works, which is still not everywhere.

This issue is still open because any other I/O exception still triggers the silent assume-they-meant-a-literal-string-with-`*`-in-it behavior. I'll rename for clarity.

**Comment by danmoseley** [comment]
cc @JeremyKuhne as both wildcards and max path are his thing.

**Comment by jasonswearingen** [comment]
This triggered for me today, seems suspiciously timed after installing a windows preview build...    Anyway, I tried all sorts of things including re-installing VS, reverted to week old commits, etc and I still got the same error.  I thought something about windows itself was corrupted, but searching hinted as this problem.

I don't know what caused this to trigger.  reverting to old commits still has the same problem, and shortening my path from  `V:\r-cleartrix\cleartrix\apps\cleartrix-dotnet\external-repo\NotNot-MonoRepo\src` to `V:\rc\c\apps\cleartrix-dotnet\external-repo\NotNot-MonoRepo\src` **still had** the same error.  

I even tried rolling back the system updates, reinstalling .NET, and even enabling long file paths in windows via regedit/gpc

only when I moved that project to `V:\rc\NotNot-MonoRepo\src` did the error go away.    

This seems very bad, because it seems that a system update triggered this and only a hugely truncated path fixes it.

Also, very very bad that the error is so obtuse.


EDIT:  it seems that me using WSL causes some file under my gitrepo to be permission denied on windows.   This is the cause of the obtuse `**/*.resx` error.    After I re-checked out my repo, the error went away.

**Comment by CEbbinghaus** [comment]
I am also encountering this error when trying to move our main codebase to utilize traversal projects. In our case there is a project that is creating an invalid junction link causing the globbing to crash when trying to enter that directory. The problem here is the lack of error rather than it not working since I expect it to crash when hitting an invalid junction rather than silently fail.

Is it possible to add failure messages rather than just silently reverting to trying a literal argument?
